src/Console/CommandScanner.php

```diff
@@ -36,30 +36,46 @@ class CommandScanner
     public function scanAll()
     {
         $shellList = [];
-
         $appNamespace = Configure::read('App.namespace');
-        $shellList['app'] = $this->scanDir(
+
+        $coreShells = $this->scanDir(
```

Comment:
The method has become a little long.

How about:

```` php
    /**
     * Scan CakePHP core, the application and plugins for shell classes
     *
     * @return array A list of shell classes divided into "CORE", "app" and "plugins".
     */
    public function scanAll()
    {
        $shellClassList = [];

        $shellClassList['CORE'] = $this->_scanCore();
        $shellClassList['app'] = $this->_scanApp();;
        $shellClassList['plugins'] = $this->_scanPlugins();

        return $shellClassList;
    }

    /**
     * Scan CakePHP core for shell classes
     *
     * @return array A list of shell classes from the CakePHP core.
     */
    protected function _scanCore()
    {
        $coreShells = $this->scanDir(
            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Shell' . DIRECTORY_SEPARATOR,
            'Cake\Shell\\',
            '',
            ['command_list']
        );
        $coreCommands = $this->scanDir(
            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Command' . DIRECTORY_SEPARATOR,
            'Cake\Command\\',
            '',
            ['command_list']
        );
        $coreShellClassList = array_merge($coreShells, $coreCommands);

        return $coreShellClassList;
    }

    /**
     * Scan CakePHP app for shell classes
     *
     * @return array A list of shell classes from the app.
     */
    protected function _scanApp()
    {
        $appNamespace = Configure::read('App.namespace');

        $appShellsCLasses = $this->scanDir(
            App::path('Shell')[0],
            $appNamespace . '\Shell\\',
            '',
            ['app']
        );
        $appCommandsClasses = $this->scanDir(
            App::path('Command')[0],
            $appNamespace . '\Command\\',
            '',
            ['app']
        );
        $appShellClassList = array_merge($appShellsCLasses, $appCommandsClasses);

        return $appShellClassList;
    }

    /**
     * Scan CakePHP plugins for shell classes
     *
     * @return array A list of shell classes from the plugins.
     */
    protected function _scanPlugins()
    {
        $pluginsShellClassList = [];
        foreach (Plugin::loaded() as $plugin) {
            $path = Plugin::classPath($plugin);
            $namespace = str_replace('/', '\\', $plugin);
            $prefix = Inflector::underscore($plugin) . '.';
            $commands = $this->scanDir($path . 'Command', $namespace . '\Command\\', $prefix, []);
            $shells = $this->scanDir($path . 'Shell', $namespace . '\Shell\\', $prefix, []);
            $pluginsShellClassList[$plugin] = array_merge($shells, $commands);
        }

        return $pluginsShellClassList;
    }
````

Reply:
Good idea!